### PR TITLE
[2022.1] Add Shadow and Additional light variant stripping

### PR DESCRIPTION
**Pull request by lukaschod**
### Purpose of this PR
This is follow up of https://github.com/Unity-Technologies/Graphics/pull/5060.

Stripping disabled variants if feature is enabled:
- Main light shadows.
- Additional light.
- Additional light shadows.

---
### Testing status
- Local GFX foundation passing.
- Tested empty project, urp template.

Empty project with most features enabled lit shader:
Before: 93
After 48

Foundation:
Before: 44k (10k lit)
After: 24k (5k lit)

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

File: com.unity.render-pipelines.universal/Shaders/Particles/ParticlesSimpleLit.shader [hunk 1]
```diff
@@ -177,7 +177,7 @@ Shader "Universal Render Pipeline/Particles/Simple Lit"
             // Universal Pipeline keywords
             #pragma multi_compile _ _MAIN_LIGHT_SHADOWS _MAIN_LIGHT_SHADOWS_CASCADE _MAIN_LIGHT_SHADOWS_SCREEN
             //#pragma multi_compile _ _ADDITIONAL_LIGHTS_VERTEX _ADDITIONAL_LIGHTS
-            #pragma multi_compile_fragment _ _ADDITIONAL_LIGHT_SHADOWS
+            //#pragma multi_compile_fragment _ _ADDITIONAL_LIGHT_SHADOWS
```
Comment (lukaschod) [inline com.unity.render-pipelines.universal/Shaders/Particles/ParticlesSimpleLit.shader:180]: @kaychang-unity is this okay?

File: com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs [hunk 2]
```diff
@@ -8,7 +8,7 @@ namespace UnityEngine.Rendering.Universal.Internal
     /// </summary>
     public class MainLightShadowCasterPass : ScriptableRenderPass
     {
-        private static class MainLightShadowConstantBuffer
+        public static class MainLightShadowConstantBuffer
```
Comment (lukaschod) [inline com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs:11]: will fix it

File: com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs [hunk 3]
```diff
@@ -48,6 +48,7 @@ enum ShaderFeatures
         MainLightShadowsCascade = (1 << 26),
         DrawProcedural = (1 << 27),
         ScreenSpaceOcclusionAfterOpaque = (1 << 28),
+        ShadowsKeepOfVariant = (1 << 29),
```
Comment (ellioman) [inline com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs:51]:
```suggestion
        ShadowsKeepOffVariant = (1 << 29),
```

Comment (ellioman) [reply com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs:51]: We need to start thinking about that we're quickly running out of bits here...

File: com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs [hunk 4]
```diff
@@ -786,10 +788,19 @@ public bool Setup(ref RenderingData renderingData)
             m_AdditionalLightsShadowmapTexture = ShadowUtils.GetTemporaryShadowTexture(renderTargetWidth, renderTargetHeight, k_ShadowmapBufferBits);
             m_MaxShadowDistanceSq = renderingData.cameraData.maxShadowDistance * renderingData.cameraData.maxShadowDistance;
             m_CascadeBorder = renderingData.shadowData.mainLightShadowCascadeBorder;
+            m_CreateEmptyShadowmap = false;
+            useNativeRenderPass = true;
 
             return true;
         }
 
+        internal void SetupEmpty()
```
Comment (ellioman) [inline com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs:797]: Can we improve the naming of this function to be a bit more descriptive?

File: com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs [hunk 5]
```diff
@@ -832,6 +849,25 @@ void Clear()
             m_AdditionalLightsShadowmapTexture = null;
         }
 
+        void EmptyAdditionalShadowmapAtlas(ref ScriptableRenderContext context)
```
Comment (ellioman) [inline com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs:852]: Can we also improve this function name as well?
Something like:
```suggestion
        void SetEmptyAdditionalShadowmapAtlas(ref ScriptableRenderContext context)
```

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs [hunk 6]
```diff
@@ -491,6 +491,8 @@ static class RenderPassBlock
 
         internal bool useDepthPriming { get; set; } = false;
 
+        internal bool stripShadowsOffVariants { get; set; } = false;
```
Comment (ellioman) [inline com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:494]: Same question here as above, is this only relevant to shadows or can we remove that prefix?

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderer.cs [hunk 7]
```diff
@@ -177,6 +178,7 @@ public UniversalRenderer(UniversalRendererData data) : base(data)
             // we inject the builtin passes in the before events.
             m_MainLightShadowCasterPass = new MainLightShadowCasterPass(RenderPassEvent.BeforeRenderingShadows);
             m_AdditionalLightsShadowCasterPass = new AdditionalLightsShadowCasterPass(RenderPassEvent.BeforeRenderingShadows);
+            this.stripShadowsOffVariants = true;
```
Comment (ellioman) [inline com.unity.render-pipelines.universal/Runtime/UniversalRenderer.cs:181]: This should be used to init the `forwardInitParams.additionalLightsAlwaysEnabled` parameter, so they stay synced up.

File: com.unity.testing.urp/Scripts/Runtime/CustomRenderPipeline/CustomRenderer.cs [hunk 8]
```diff
@@ -4,14 +4,21 @@ namespace UnityEngine.Rendering.Universal
 {
     public class CustomRenderer : ScriptableRenderer
```
Comment (ellioman) [inline com.unity.testing.urp/Scripts/Runtime/CustomRenderPipeline/CustomRenderer.cs:5]: Does this test make sense?
With this change we're now using internal functionality not available to users.

**Review by ellioman — CHANGES_REQUESTED** [review]
Added a few comments that I want you to take a look at before approving.

File: com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs (hunk 3 above)
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs:51]: This is for shadow and additional light keywords. It needs somekind of prefix to indicate that it does not control for all features

File: com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs (hunk 5 above)
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs:852]: I will add Execute as prefix, setup might be confusing with our existing setup functions

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs (hunk 6 above)
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:494]: yeap only shadows and additional lights

File: com.unity.testing.urp/Scripts/Runtime/CustomRenderPipeline/CustomRenderer.cs (hunk 8 above)
Comment (lukaschod) [reply com.unity.testing.urp/Scripts/Runtime/CustomRenderPipeline/CustomRenderer.cs:5]: That is good question, I personally would prefer if we had this test as separate project. Because not using this interal function, will result in lots of variants that we do not need

File: com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs (hunk 5 above)
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs:852]: Of course render would be more consistant, but might be confussing as there is not rendering :/. Naming is hard...

File: com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs (hunk 3 above)
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs:51]: Yeap this needs to be struct in my opnion, but for time being I think it can be 64bit too

File: com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs (hunk 4 above)
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs:797]: SetupForEmptyRendering? Quite long :D

File: com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs (hunk 3 above)
Comment (pbbastian) [reply com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs:51]: Agree, we just felt like bringing the bit issue up, but not that it should be fixed in this PR. And yeah, I agree with struct being a good choice instead.

Regarding the name, could it be something like `LightingKeepOffVariants`? IMO that's covers shadows as well. Maybe with a comment somewhere about what it exactly covers.

File: com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs (hunk 5 above)
Comment (pbbastian) [reply com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs:852]: I think `Set` is different enough from `Setup` prefix

File: com.unity.testing.urp/Scripts/Runtime/CustomRenderPipeline/CustomRenderer.cs (hunk 8 above)
Comment (pbbastian) [reply com.unity.testing.urp/Scripts/Runtime/CustomRenderPipeline/CustomRenderer.cs:5]: ohhh that is a good point. Then I think we should just do what you've done here actually.

File: com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs (hunk 3 above)
Comment (pbbastian) [reply com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs:51]: or maybe just go all in and call it `KeepOffVariantsShadowsAdditionalLights`, although that is a bit long 😅

**Review by ellioman — APPROVED** [review]

File: com.unity.render-pipelines.universal/Runtime/ForwardLights.cs [hunk 9]
```diff
@@ -353,10 +359,11 @@ public void Setup(ScriptableRenderContext context, ref RenderingData renderingDa
 
                 SetupShaderLightConstants(cmd, ref renderingData);
 
+                bool lightCountCheck = (m_AdditionalLightsAlwaysEnabled && renderingData.lightData.maxPerObjectAdditionalLightsCount > 0) || additionalLightsCount > 0;
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/ForwardLights.cs:362]: can `additionalLightsCount` be changed at runtime and you get into a situation that the keyword and stripped and at runtime you try to enable it?

**Review by phi-lira — APPROVED** [review]
Approving with some comments.

**Review by pbbastian — APPROVED** [review]

File: com.unity.render-pipelines.universal/Shaders/Particles/ParticlesSimpleLit.shader (hunk 1 above)
Comment (kaychang-unity) [reply com.unity.render-pipelines.universal/Shaders/Particles/ParticlesSimpleLit.shader:180]: I think you should keep `_ADDITIONAL_LIGHT_SHADOWS` variant to allow SimpleLit shaders to to generate shadows for additional lights in deferred.

Comment (lukaschod) [reply com.unity.render-pipelines.universal/Shaders/Particles/ParticlesSimpleLit.shader:180]: awesome that is what I assumed

File: com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs [hunk 10]
```diff
@@ -832,6 +849,25 @@ void Clear()
...
+            cmd.SetGlobalTexture(m_AdditionalLightsShadowmap.id, Texture2D.whiteTexture);
+            if (RenderingUtils.useStructuredBuffer)
+            {
+                var shadowParamsBuffer = ShaderData.instance.GetAdditionalLightShadowParamsStructuredBuffer(m_AdditionalLightIndexToShadowParams.Length);
+                shadowParamsBuffer.SetData(m_AdditionalLightIndexToShadowParams);
+                cmd.SetGlobalBuffer(m_AdditionalShadowParams_SSBO, shadowParamsBuffer);
+            }
+            else
+            {
+                cmd.SetGlobalVectorArray(AdditionalShadowsConstantBuffer._AdditionalShadowParams, m_AdditionalLightIndexToShadowParams);
```
[the remark this answers is not in the document]
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs:876]: fixed

File: com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs [hunk 11]
```diff
@@ -114,10 +116,19 @@ public bool Setup(ref RenderingData renderingData)
...
 
             return true;
         }
 
+        internal void SetupForEmptyRendering()
+        {
+            m_MainLightShadowmapTexture = ShadowUtils.GetTemporaryShadowTexture(1, 1, k_ShadowmapBufferBits);
+            m_CreateEmptyShadowmap = true;
+            useNativeRenderPass = false;
+        }
```
[the remark this answers is not in the document]
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs]: fixed

File: com.unity.render-pipelines.universal/Runtime/ForwardLights.cs (hunk 9 above)
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Runtime/ForwardLights.cs]: add additional check to prevent this

**Review by aleksandrasdzikia — APPROVED** [review]
Looks good. Tested URP template and Viking Village. Visual correctness looks good, no performance regression. Played around with different settings, didn't find any issues.

**Comment by lukaschod** [comment]
URP_Foundation on Android_OpenGLES3_Standalone_il2cpp_Linear on version trunk 
https://unity-ci.cds.internal.unity3d.com/job/9183640
URP_Foundation on Android_Vulkan_Standalone_il2cpp_Linear on version trunk
https://unity-ci.cds.internal.unity3d.com/job/9167131
URP_Lighting on Android_OpenGLES3_Standalone_il2cpp_Linear on version trunk
https://unity-ci.cds.internal.unity3d.com/job/9167152
URP_Lighting on Android_Vulkan_Standalone_il2cpp_Linear on version trunk
Connection Failure on master

URP_PostPro on Android_Vulkan_Standalone_il2cpp_Linear on version trunk
https://unity-ci.cds.internal.unity3d.com/job/9156314
URP_PostPro on Linux_Vulkan_Standalone_mono_Linear on version trunk
https://unity-ci.cds.internal.unity3d.com/job/9156313
URP_PostPro on Win_Vulkan_Standalone_mono_Linear on version trunk
failing tests on master
UniversalGraphicsTests.145_SoftParticlesStackWithPP
UniversalGraphicsTests.145_SoftParticlesStackWithPP_deferred

URP_PostPro on OSX_Metal_playmode_mono_Linear on version trunk
https://unity-ci.cds.internal.unity3d.com/job/9173533
Some random failure on master

URP_PostPro on Win_DX11_playmode_XR_mono_Linear on version trunk
https://unity-ci.cds.internal.unity3d.com/job/9149555
Same issue on master
One or more child steps have failed.
Editor exited with non-zero exit code: -1073741819

URP_Terrain on Android_OpenGLES3_Standalone_il2cpp_Linear on version trunk
https://unity-ci.cds.internal.unity3d.com/job/9159438
URP_Terrain on Android_Vulkan_Standalone_il2cpp_Linear on version trunk 
https://unity-ci.cds.internal.unity3d.com/job/9167173
URP_Terrain on OSX_Metal_playmode_mono_Linear on version trunk
Same issue on master

Universal_Stereo on Win__Standalone_mono_Linear on version trunk
https://unity-ci.cds.internal.unity3d.com/job/9183646
master failing LWGraphicsTests.126_SampleDepth

**Review by phi-lira — APPROVED** [review]
LGTM.
